tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
 
         // To preserve this class's invariants, we can't just clear the buffer; we also need to refill the
         // buffer from disk.
-        if (this.diskPos != this.lo) {
-            super.seek(this.lo);
-            this.diskPos = this.lo;
-        }
-        this.hi = this.lo + fillBuffer();
+        fillBuffer();
     }
-    
-    /* overrides RandomAccessFile.close() */
+
+    @Override
     public void close() throws IOException {
-        // Assert.check(!this.closed);
-        this.flush();
-        // SZ Feb 24, 2009: never read locally
-        // this.closed = true;
-        synchronized (mu) {
-            // grow "availBuffs" array if necessary
-            if (numAvailBuffs >= availBuffs.length) {
-                byte[][] newBuffs = new byte[numAvailBuffs + 10][];
-                System.arraycopy(availBuffs, 0, newBuffs, 0, numAvailBuffs);
-                availBuffs = newBuffs;
+        try {
+            // Assert.check(!this.closed);
```

Comment:
currently, the implementation allows read after close. We close the underlying file, give back the buffer to the pool but the buffer is not nullified, and subsequent reads that don't access the file will succeed.

Sample test adopted from the testRead (see last three rows)
```
	@Test
	public void testRead() throws IOException {
		final File tmpFile = File.createTempFile("BufferedRandomAccessFileTest_testRead", ".bin");
		tmpFile.deleteOnExit();
		final java.io.RandomAccessFile raf = new BufferedRandomAccessFile(tmpFile, "rw");
		for (long i = 0L; i < BufferedRandomAccessFile.BuffSz / 8; i++) {
			raf.writeLong(i);
		}

		raf.seek(0);
		for (long i = 0l; i < BufferedRandomAccessFile.BuffSz / 8; i++) {
			assertEquals(i, raf.readLong());
		}
                raf.seek(0); //seek back to zero
		raf.close(); // closed
                assertEquals(0, raf.readLong()); // still works
	}
```
This could be a problem in case the buffer is reassigned to a different BRAF. To reduce the state, it would be sufficient to call "isClosed" on the underlinying RAF. Because there is no isClosed method, getFilePointer could be used for its side effect of failing if the underlying object is closed.

Reply:
I have mixed feelings about this.

> the implementation allows read after close

Yes and no. The implementation does not throw an exception if you do a read after close, so in that sense it "allows" read after close. But, the implementation also doesn't specify behavior if you read after close, so in that sense it "disallows" read after close. In other words, _clients_ are responsible for ensuring that they do not interact with closed files.

In this case, I would peronsally rather push responsibility to the callers---but it's true that the official [RandomAccessFile contract](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/RandomAccessFile.html) _does_ require `IOException` to be thrown for _some_ read and write methods if the file is closed (but weirdly, not all of them). In your specific example there is no specified behavior, and therefore the current behavior ("return an arbitrary long without throwing an exception") is legal. But if you replace the final `readLong()` with `read(byte[])`, it is indeed a demonstration of a bug, since that method requires an exception if the file is closed.

Also, because you called this out I noticed that this `close()` method is not idempotent, and [it must be since this class implements `Closeable`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/AutoCloseable.html). A follow-up call will return `buff` to the `availBuffs` pool a second time, which could be catastrophic.

In any case, I think there is a strong case for reinstating the commented-out `closed` field, which would make it very easy to fix these problems. I will look into doing this on Monday.

Reply:
Aside: it is also irksome to me that the official contract for `RandomAccessFile` is to throw `IOException` if the file is closed, when [`IllegalStateException`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/IllegalStateException.html) already exists and was designed for _exactly this purpose_:

> Signals that a method has been invoked at an illegal or inappropriate time.

But, to my surprise, `RandomAccessFile` is _older_ than `IllegalStateException`, so I suppose it makes sense.

Reply:
@FedericoPonzi @lemmy On second thought, I do not plan to address this comment in this PR. There is no _new_ bug here, so I think it makes sense to address this in a separate PR.